File: nodejs_implementation/src/utils/file-io.js
/**
 * File I/O Utility for DeepWiki to Markdown Converter
 * 
 * This module provides file I/O functionality for the DeepWiki to Markdown converter.
 */

const fs = require('fs');
const path = require('path');
const yaml = require('js-yaml');
const { logger } = require('./logger');

/**
 * Save content to a Markdown file.
 * 
 * @param {string} filePath - Path to save the file
 * @param {string} content - Content to save
 * @returns {boolean} True if successful, false otherwise
 */
function saveMarkdownFile(filePath, content) {
  try {
    // Create directory if it doesn't exist
    const dirPath = path.dirname(filePath);
    fs.mkdirSync(dirPath, { recursive: true });
    
    // Save the file
    fs.writeFileSync(filePath, content, 'utf8');
    
    logger.debug(`Saved Markdown file: ${filePath}`);
    return true;
  } catch (error) {
    logger.error(`Error saving Markdown file ${filePath}: ${error.message}`);
    return false;
  }
}

/**
 * Read content from a Markdown file.
 * 
 * @param {string} filePath - Path to the file
 * @returns {string|null} File content or null if an error occurred
 */
function readMarkdownFile(filePath) {
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    
    logger.debug(`Read Markdown file: ${filePath}`);
    return content;
  } catch (error) {
    logger.error(`Error reading Markdown file ${filePath}: ${error.message}`);
    return null;
  }
}

/**
 * Save data to a JSON file.
 * 
 * @param {string} filePath - Path to save the file
 * @param {Object} data - Data to save
 * @returns {boolean} True if successful, false otherwise
 */
function saveJsonFile(filePath, data) {
  try {
    // Create directory if it doesn't exist
    const dirPath = path.dirname(filePath);
    fs.mkdirSync(dirPath, { recursive: true });
    
    // Save the file
    fs.writeFileSync(filePath, JSON.stringify(data, null, 2), 'utf8');
    
    logger.debug(`Saved JSON file: ${filePath}`);
    return true;
  } catch (error) {
    logger.error(`Error saving JSON file ${filePath}: ${error.message}`);
    return false;
  }
}

/**
 * Read data from a JSON file.
 * 
 * @param {string} filePath - Path to the file
 * @returns {Object|null} File data or null if an error occurred
 */
function readJsonFile(filePath) {
  try {
    const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    
    logger.debug(`Read JSON file: ${filePath}`);
    return data;
  } catch (error) {
    logger.error(`Error reading JSON file ${filePath}: ${error.message}`);
    return null;
  }
}

/**
 * Save data to a YAML file.
 * 
 * @param {string} filePath - Path to save the file
 * @param {Object} data - Data to save
 * @returns {boolean} True if successful, false otherwise
 */
function saveYamlFile(filePath, data) {
  try {
    // Create directory if it doesn't exist
    const dirPath = path.dirname(filePath);
    fs.mkdirSync(dirPath, { recursive: true });
    
    // Save the file
    const yamlStr = yaml.dump(data, {
      indent: 2,
      lineWidth: -1,
      noRefs: true,
      sortKeys: false
    });
    fs.writeFileSync(filePath, yamlStr, 'utf8');
    
    logger.debug(`Saved YAML file: ${filePath}`);
    return true;
  } catch (error) {
    logger.error(`Error saving YAML file ${filePath}: ${error.message}`);
    return false;
  }
}

/**
 * Read data from a YAML file.
 * 
 * @param {string} filePath - Path to the file
 * @returns {Object|null} File data or null if an error occurred
 */
function readYamlFile(filePath) {
  try {
    const data = yaml.load(fs.readFileSync(filePath, 'utf8'));
    
    logger.debug(`Read YAML file: ${filePath}`);
    return data;
  } catch (error) {
    logger.error(`Error reading YAML file ${filePath}: ${error.message}`);
    return null;
  }
}

/**
 * Check if a file exists.
 * 
 * @param {string} filePath - Path to the file
 * @returns {boolean} True if the file exists, false otherwise
 */
function fileExists(filePath) {
  try {
    return fs.existsSync(filePath);
  } catch (error) {
    logger.error(`Error checking if file exists ${filePath}: ${error.message}`);
    return false;
  }
}

/**
 * List files in a directory.
 * 
 * @param {string} dirPath - Path to the directory
 * @param {Object} [options={}] - Options
 * @param {boolean} [options.recursive=false] - Whether to list files recursively
 * @param {string} [options.extension=null] - Filter by file extension
 * @returns {Array<string>} List of file paths
 */
function listFiles(dirPath, options = {}) {
  const { recursive = false, extension = null } = options;
  
  try {
    if (!fs.existsSync(dirPath)) {
      logger.warn(`Directory does not exist: ${dirPath}`);
      return [];
    }
    
    if (!fs.statSync(dirPath).isDirectory()) {
      logger.warn(`Path is not a directory: ${dirPath}`);
      return [];
    }
    
    let files = [];
    
    const items = fs.readdirSync(dirPath);
    
    for (const item of items) {
      const itemPath = path.join(dirPath, item);
      const stat = fs.statSync(itemPath);
      
      if (stat.isDirectory() && recursive) {
        files = files.concat(listFiles(itemPath, options));
      } else if (stat.isFile()) {
        if (!extension || path.extname(itemPath) === extension) {
          files.push(itemPath);
        }
      }
    }
    
    return files;
  } catch (error) {
    logger.error(`Error listing files in ${dirPath}: ${error.message}`);
    return [];
  }
}

module.exports = {
  saveMarkdownFile,
  readMarkdownFile,
  saveJsonFile,
  readJsonFile,
  saveYamlFile,
  readYamlFile,
  fileExists,
  listFiles
};

File: deepwiki_to_md/utils/file_io.py
"""
File I/O Utilities for DeepWiki to Markdown Converter

This module provides file I/O utility functions for the DeepWiki to Markdown converter.
"""

import os
import logging
from typing import Optional, Dict, Any
import json
import yaml

logger = logging.getLogger(__name__)


def save_markdown_file(file_path: str, content: str) -> bool:
    """
    Save content to a Markdown file.
    
    Args:
        file_path (str): Path to save the file
        content (str): Content to save
        
    Returns:
        bool: True if successful, False otherwise
    """
    try:
        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        # Save the file
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)

        logger.debug(f"Saved Markdown file: {file_path}")
        return True

    except Exception as e:
        logger.error(f"Error saving Markdown file {file_path}: {str(e)}")
        return False


def read_markdown_file(file_path: str) -> Optional[str]:
    """
    Read content from a Markdown file.
    
    Args:
        file_path (str): Path to the file
        
    Returns:
        Optional[str]: File content or None if an error occurred
    """
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()

        logger.debug(f"Read Markdown file: {file_path}")
        return content

    except Exception as e:
        logger.error(f"Error reading Markdown file {file_path}: {str(e)}")
        return None


def save_json_file(file_path: str, data: Dict[str, Any]) -> bool:
    """
    Save data to a JSON file.
    
    Args:
        file_path (str): Path to save the file
        data (Dict[str, Any]): Data to save
        
    Returns:
        bool: True if successful, False otherwise
    """
    try:
        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        # Save the file
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, ensure_ascii=False, indent=2)

        logger.debug(f"Saved JSON file: {file_path}")
        return True

    except Exception as e:
        logger.error(f"Error saving JSON file {file_path}: {str(e)}")
        return False


def read_json_file(file_path: str) -> Optional[Dict[str, Any]]:
    """
    Read data from a JSON file.
    
    Args:
        file_path (str): Path to the file
        
    Returns:
        Optional[Dict[str, Any]]: File data or None if an error occurred
    """
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)

        logger.debug(f"Read JSON file: {file_path}")
        return data

    except Exception as e:
        logger.error(f"Error reading JSON file {file_path}: {str(e)}")
        return None


def save_yaml_file(file_path: str, data: Dict[str, Any]) -> bool:
    """
    Save data to a YAML file.
    
    Args:
        file_path (str): Path to save the file
        data (Dict[str, Any]): Data to save
        
    Returns:
        bool: True if successful, False otherwise
    """
    try:
        # Create directory if it doesn't exist
        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        # Save the file
        with open(file_path, 'w', encoding='utf-8') as f:
            yaml.dump(data, f, default_flow_style=False, allow_unicode=True)

        logger.debug(f"Saved YAML file: {file_path}")
        return True

    except Exception as e:
        logger.error(f"Error saving YAML file {file_path}: {str(e)}")
        return False


def read_yaml_file(file_path: str) -> Optional[Dict[str, Any]]:
    """
    Read data from a YAML file.
    
    Args:
        file_path (str): Path to the file
        
    Returns:
        Optional[Dict[str, Any]]: File data or None if an error occurred
    """
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)

        logger.debug(f"Read YAML file: {file_path}")
        return data

    except Exception as e:
        logger.error(f"Error reading YAML file {file_path}: {str(e)}")
        return None
